src/main/kotlin/com/hypto/iam/server/service/OrganizationsService.kt

```diff
@@ -109,6 +115,84 @@ class OrganizationsServiceImpl : KoinComponent, OrganizationsService {
         }
     }
 
+    override suspend fun createOauthOrganization(
+        companyName: String,
+        name: String,
+        email: String
+    ): Pair<Organization, TokenResponse> {
+        val organizationId = idGenerator.organizationId()
+        val username = idGenerator.username()
+
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            return txMan.wrap {
+                // Create Organization
+                organizationRepo.insert(
+                    Organizations(
+                        organizationId,
+                        companyName,
+                        "",
+                        ResourceHrn(
+                            organization = organizationId,
+                            resource = IamResources.USER,
+                            resourceInstance = username
+                        ).toString(),
+                        null,
+                        LocalDateTime.now(),
+                        LocalDateTime.now()
+                    )
+                )
+
+                // Create root user for the organization
+                val rootUser = usersService.createOauthUser(
+                    organizationId = organizationId,
+                    username = username,
+                    preferredUsername = null,
+                    name = name,
+                    email = email,
+                    createdBy = "iam-system",
+                    verified = true,
+                    loginAccess = true
+                )
+
+                val policyHrns = policyTemplatesService
+                    .createPersistAndReturnRootPolicyRecordsForOrganization(organizationId, rootUser)
+                    .map { ResourceHrn(it.hrn) }
+
+                // TODO: Avoid this duplicate call be returning the created organization from `organizationRepo.insert`
+                val organization = getOrganization(organizationId)
+                val userHrn = hrnFactory.getHrn(rootUser.hrn)
+
+                if (policyHrns.isNotEmpty()) {
+                    principalPolicyService.attachPoliciesToUser(userHrn, policyHrns)
+                }
+
+                val token = tokenService.generateJwtToken(userHrn)
+
+                if (AppConfig.configuration.postHook.signUp != null) {
```

Comment:
What is the expected behaviour in case of failures?
Case-1. postHook call succeeds but transaction commit fails => IAM won't have org but upstream will have recorded it.
  
Case-2. postHook call fails and so transaction gets rolledback => Upstream wouln't have recorded it but IAM would have the organization and user.

Either way, the same person cannot create a new account when email_id uniqueness is enabled.

Reply:
Discussed in call that:

Case-1: Will have stale record upstream, but no record in IAM.
Case-2: No record in both upstream and IAM

So, in effect the end user won't be affected.